openmeter/billing/service/lineservice/usagebasedline_test.go

```diff
@@ -0,0 +1,1366 @@
+				Tiers:         testTiers,
+				MaximumAmount: lo.ToPtr(alpacadecimal.NewFromFloat(125)),
+			}),
+			lineMode: lastInPeriodSplitLineMode,
+			usage: featureUsageResponse{
+				LinePeriodQty: alpacadecimal.NewFromFloat(7),
+			},
+			expect: newDetailedLinesInput{
+				{
+					Name:                   "feature: flat price for tier 2",
+					Amount:                 alpacadecimal.NewFromFloat(150),
+					Quantity:               alpacadecimal.NewFromFloat(1),
+					ChildUniqueReferenceID: GraduatedFlatPriceChildUniqueReferenceID,
+					PaymentTerm:            plan.InArrearsPaymentTerm,
+					Discounts: []billingentity.LineDiscount{
+						{
+							Description: lo.ToPtr("Maximum spend discount for charges over 125"),
+							Amount:      alpacadecimal.NewFromFloat(25),
+							Type:        lo.ToPtr(billingentity.MaximumSpendLineDiscountType),
+							Source:      billingentity.CalculatedLineDiscountSource,
+						},
+					},
+				},
+			},
+		})
+	})
+}
+
+func TestTieredVolumeCalculation(t *testing.T) {
+	testTiers := []plan.PriceTier{
+		{
+			UpToAmount: lo.ToPtr(alpacadecimal.NewFromFloat(5)),
+			FlatPrice: &plan.PriceTierFlatPrice{
+				// 20/unit
+				Amount: alpacadecimal.NewFromFloat(100),
+			},
+		},
+		{
+			UpToAmount: lo.ToPtr(alpacadecimal.NewFromFloat(10)),
+			FlatPrice: &plan.PriceTierFlatPrice{
+				// 10/unit
+				Amount: alpacadecimal.NewFromFloat(50),
+			},
+		},
+		{
+			UpToAmount: lo.ToPtr(alpacadecimal.NewFromFloat(15)),
+			UnitPrice: &plan.PriceTierUnitPrice{
+				Amount: alpacadecimal.NewFromFloat(5),
+			},
+		},
+		{
+			UnitPrice: &plan.PriceTierUnitPrice{
+				Amount: alpacadecimal.NewFromFloat(1),
+			},
+		},
+	}
+
+	t.Run("tiered volume, mid price, flat only => no lines are output", func(t *testing.T) {
+		runUBPTest(t, ubpCalculationTestCase{
+			price: plan.NewPriceFrom(plan.TieredPrice{
+				Mode:  plan.VolumeTieredPrice,
+				Tiers: testTiers,
+			}),
+			lineMode: midPeriodSplitLineMode,
+			usage: featureUsageResponse{
+				PreLinePeriodQty: alpacadecimal.NewFromFloat(7),
+				LinePeriodQty:    alpacadecimal.NewFromFloat(1),
+			},
+			expect: newDetailedLinesInput{},
+		})
+	})
+
+	t.Run("tiered volume, last price, no usage", func(t *testing.T) {
+		runUBPTest(t, ubpCalculationTestCase{
+			price: plan.NewPriceFrom(plan.TieredPrice{
+				Mode:  plan.VolumeTieredPrice,
+				Tiers: testTiers,
+			}),
+			lineMode: lastInPeriodSplitLineMode,
+			usage: featureUsageResponse{
+				LinePeriodQty: alpacadecimal.NewFromFloat(0),
+			},
+			expect: newDetailedLinesInput{},
+		})
+	})
+
+	t.Run("tiered volume, single period multiple tier usage", func(t *testing.T) {
+		runUBPTest(t, ubpCalculationTestCase{
+			price: plan.NewPriceFrom(plan.TieredPrice{
+				Mode:  plan.VolumeTieredPrice,
+				Tiers: testTiers,
+			}),
+			lineMode: singlePerPeriodLineMode,
+			usage: featureUsageResponse{
+				LinePeriodQty: alpacadecimal.NewFromFloat(22),
+			},
+			expect: newDetailedLinesInput{
```

Comment:
on qty 22 we should have only one line item on the highest discount, see:
https://openmeter.io/blog/tiered-pricing-model#example-comparison

Reply:
This was a missmatch in naming, fixed.